src/mptcpize.c

```diff
@@ -0,0 +1,226 @@
+		"\t\t\t forcing MPTCP socket usage instead of TCP\n"
+		"\t\t\t if the '-d' argument is provided, dump messages on stderr\n"
+		"\t\t\t when a TCP socket is forced to MPTCP\n"
+		"\tservice <unit>       create/update <unit>_mptcp systemd unit, which\n"
+		"\t\t\t force the given service to run under the above launcher\n"
+		"\thelp                 show this help\n");
+}
+
+#define MPTCPWRAP_ENV "LD_PRELOAD="PKGLIBDIR"/libmptcpwrap.so.0.0."LIBREVISION
+
+static int run(int argc, char *av[])
+{
+	int i, nr = 0, debug = 0;
+	char **envp, **argv;
+
+	if (argc > 0 && !strcmp(av[0], "-d")) {
+		debug = 1;
+		argc--;
+		av++;
+	}
+
+	if (argc < 1) {
+		fprintf(stderr, "missing command argument");
+		help();
+		return -1;
+	}
+
+	// build environment, copying the current one ...
+	while (environ[nr])
+		nr++;
+	envp = calloc(nr + 3, sizeof(char *));
+	if (!envp)
+		error(1, errno, "can't allocate env list");
+
+	// ... filtering out any 'LD_PRELOAD' ...
+	nr = 0;
+	i = 0;
+	while (environ[nr]) {
+		if (strncmp(environ[nr], "LD_PRELOAD=", 11)) {
+			envp[i] = environ[nr];
+			i++;
+		}
+		nr++;
+	}
+
+	// ... appending the mptcpwrap preload...
+	envp[i++] = MPTCPWRAP_ENV;
+
+	// ... and enable dbg if needed
+	if (debug)
+		envp[i++] = "MPTCPWRAP_DEBUG=1";
+
+	// build the NULL terminated arg list
+	argv = calloc(argc + 1, sizeof(char *));
+	if (!argv)
+		error(1, errno, "can't allocate argument list");
+
+	memcpy(argv, av, argc * sizeof(char*));
+	return execvpe(argv[0], argv, envp);
+}
+
+#define SYSTEMCTL_SHOW "systemctl show -p FragmentPath "
+
+static char *locate_unit(const char *name)
+{
+	char *cmd, *line = NULL;
+	FILE *systemctl;
+	size_t len = 0;
+	ssize_t read;
+
+	if (!access(name, R_OK))
+		return strdup(name);
+
+	len = strlen(name) + 1 + strlen(SYSTEMCTL_SHOW);
+	cmd = malloc(len);
+	if (!cmd)
+		error(1, 0, "can't allocate systemctl command string");
+	sprintf(cmd, SYSTEMCTL_SHOW"%s", name);
+
+	systemctl = popen(cmd, "r");
+	if (!systemctl)
+		error(1, errno, "can't execute %s", cmd);
+
+	while ((read = getline(&line, &len, systemctl)) != -1) {
+		if (!strncmp(line, "FragmentPath=", 13)) {
+			char *ret = strdup(&line[13]);
+
+			// trim trailing newline, if any
+			len = strlen(ret);
+			if (len > 0 && ret[len - 1] == '\n')
+				ret[len - 1] = 0;
+			free(line);
+			pclose(systemctl);
+			return ret;
+		}
+	}
+
+	error(1, 0, "can't find FragmentPath attribute for unit %s", name);
+
+	// never reached: just silence gcc
+	return NULL;
+}
+
+static int unit_update(int argc, char *argv[], int enable)
+{
+	char *unit, *line = NULL;
+	char dst_path[PATH_MAX];
+	int append_env = enable;
+	size_t len = 0;
+	ssize_t read;
+	FILE *src;
+	int dst;
+
+	if (argc < 1) {
+		fprintf(stderr, "missing unit argument");
+		help();
+		return -1;
+	}
+
+	unit = locate_unit(argv[0]);
+	src = fopen(unit, "r");
+	if (!src)
+		error(1, errno, "can't open file %s", unit);
+
+	strcpy(dst_path, "/tmp/unit_XXXXXX");
```

Comment:
oks, I'll use 'tempnam(NULL, "unit")'

Reply:
looks like it's better to avoid tempnam() and friends, as gcc explicitly warns on such functions usage.

Additionally it looks like /tmp will always be available. So I'll revert back to  mkstemp()